File: android/app/src/main/kotlin/com/stillmoment/domain/models/MeditationTimer.kt
package com.stillmoment.domain.models

import java.util.Locale

/**
 * Domain model representing a meditation timer session.
 *
 * This is an immutable data class. All mutations return new instances.
 * Equivalent to the Swift struct MeditationTimer.
 *
 * @property durationMinutes Duration of the timer in minutes (1-60)
 * @property remainingSeconds Remaining time in seconds
 * @property state Current state of the timer
 * @property remainingPreparationSeconds Remaining preparation seconds (preparationTimeSeconds→0 before timer starts)
 * @property preparationTimeSeconds Duration of preparation in seconds (configured at initialization)
 * @property attunementDurationSeconds Duration of attunement audio in seconds (0 = no attunement)
 * @property silentPhaseStartRemaining Remaining seconds when silent meditation phase started (after attunement)
 * @property lastIntervalGongAt Remaining seconds when last interval gong was played
 */
data class MeditationTimer(
    val durationMinutes: Int,
    val remainingSeconds: Int,
    val state: TimerState,
    val remainingPreparationSeconds: Int = 0,
    val preparationTimeSeconds: Int = DEFAULT_PREPARATION_TIME,
    val attunementDurationSeconds: Int = 0,
    val silentPhaseStartRemaining: Int? = null,
    val lastIntervalGongAt: Int? = null
) {
    init {
        require(durationMinutes in 1..60) {
            "Invalid duration: $durationMinutes minutes. Duration must be between 1 and 60 minutes."
        }
    }

    /** Returns total duration in seconds */
    val totalSeconds: Int
        get() = durationMinutes * 60

    /** Returns progress as a value between 0.0 and 1.0 */
    val progress: Float
        get() =
            if (totalSeconds > 0) {
                1.0f - (remainingSeconds.toFloat() / totalSeconds.toFloat())
            } else {
                0.0f
            }

    /** Checks if timer has completed */
    val isCompleted: Boolean
        get() = remainingSeconds <= 0

    // MARK: - Computed Properties (Display)

    /** Whether currently in preparation phase */
    val isPreparation: Boolean
        get() = state == TimerState.Preparation

    /** Whether the session is active (all states between Preparation and EndGong, inclusive) */
    val isActive: Boolean
        get() = state != TimerState.Idle && state != TimerState.Completed

    /** Whether the timer is in the main running phase */
    val isRunning: Boolean
        get() = state == TimerState.Running

    /** Whether the timer can be reset (not idle) */
    val canReset: Boolean
        get() = state != TimerState.Idle

    /** Whether the timer is in idle state (ready to start) */
    val canStart: Boolean
        get() = state == TimerState.Idle

    /** Formatted time string: preparation seconds as integer, or MM:SS for other phases */
    val formattedTime: String
        get() =
            if (isPreparation) {
                "$remainingPreparationSeconds"
            } else {
                val minutes = remainingSeconds / 60
                val seconds = remainingSeconds % 60
                String.format(Locale.ROOT, "%02d:%02d", minutes, seconds)
            }

    /**
     * Returns a copy with updated remaining seconds and any domain events that occurred.
     *
     * @param intervalSettings Optional interval gong configuration. When provided,
     *   `tick()` checks if an interval gong is due and emits [TimerEvent.IntervalGongDue].
     *   When `null`, no interval gong detection occurs.
     * @return Pair of (updated timer, events that occurred during this tick)
     */
    fun tick(intervalSettings: IntervalSettings? = null): Pair<MeditationTimer, List<TimerEvent>> {
        return when (state) {
            TimerState.Preparation -> {
                val newTimer = tickPreparation()
                val events = if (newTimer.state == TimerState.StartGong) {
                    listOf(TimerEvent.PreparationCompleted)
                } else {
                    emptyList()
                }
                newTimer to events
            }
            TimerState.StartGong -> {
                val newTimer = tickRunning()
                val events = if (newTimer.state == TimerState.EndGong) {
                    listOf(TimerEvent.MeditationCompleted)
                } else {
                    emptyList()
                }
                newTimer to events
            }
            TimerState.Attunement -> {
                val newTimer = tickAttunement()
                val events = if (newTimer.state == TimerState.EndGong) {
                    listOf(TimerEvent.MeditationCompleted)
                } else {
                    emptyList()
                }
                newTimer to events
            }
            TimerState.Running -> tickRunningWithEvents(intervalSettings)
            TimerState.Idle, TimerState.EndGong, TimerState.Completed -> this to emptyList()
        }
    }

    /** Ticks the running state with interval gong detection and event emission */
    private fun tickRunningWithEvents(intervalSettings: IntervalSettings?): Pair<MeditationTimer, List<TimerEvent>> {
        val ticked = tickRunning()

        if (ticked.state == TimerState.EndGong) {
            return ticked to listOf(TimerEvent.MeditationCompleted)
        }

        if (intervalSettings != null &&
            ticked.shouldPlayIntervalGong(intervalSettings.intervalMinutes, intervalSettings.mode)
        ) {
            return ticked.markIntervalGongPlayed() to listOf(TimerEvent.IntervalGongDue)
        }

        return ticked to emptyList()
    }

    /** Ticks the preparation countdown. Transitions to StartGong when preparation finishes. */
    private fun tickPreparation(): MeditationTimer {
        val newPreparation = maxOf(0, remainingPreparationSeconds - 1)
        val newState = if (newPreparation <= 0) TimerState.StartGong else TimerState.Preparation
        return copy(
            remainingPreparationSeconds = newPreparation,
            state = newState
        )
    }

    /**
     * Ticks the attunement phase (meditation timer decrements, never auto-transitions to running).
     * The transition to Running is event-driven via endAttunement().
     */
    private fun tickAttunement(): MeditationTimer {
        val newRemaining = maxOf(0, remainingSeconds - 1)
        val newState = if (newRemaining <= 0) TimerState.EndGong else TimerState.Attunement
        return copy(
            remainingSeconds = newRemaining,
            state = newState
        )
    }

    /** Ticks the main meditation timer (used for StartGong and Running states). */
    private fun tickRunning(): MeditationTimer {
        val newRemaining = maxOf(0, remainingSeconds - 1)
        val newState = if (newRemaining <= 0) TimerState.EndGong else state
        return copy(
            remainingSeconds = newRemaining,
            state = newState
        )
    }

    /** Returns a copy with updated state */
    fun withState(newState: TimerState): MeditationTimer {
        return copy(state = newState)
    }

    /** Returns a copy ready for preparation phase (uses configured preparationTimeSeconds) */
    fun startPreparation(): MeditationTimer {
        return copy(
            state = TimerState.Preparation,
            remainingPreparationSeconds = preparationTimeSeconds,
            lastIntervalGongAt = null
        )
    }

    /**
     * Returns a copy transitioned to Attunement state.
     * Called when start gong finishes and an attunement is configured.
     */
    fun startAttunement(): MeditationTimer {
        return copy(state = TimerState.Attunement)
    }

    /**
     * Returns a copy transitioned from Attunement to Running.
     * Called when the attunement audio finishes playing (event-driven).
     * Sets silentPhaseStartRemaining to current remaining seconds for interval gong calculations.
     */
    fun endAttunement(): MeditationTimer {
        return copy(
            state = TimerState.Running,
            silentPhaseStartRemaining = remainingSeconds
        )
    }

    /** Returns a copy with updated interval gong timestamp */
    fun markIntervalGongPlayed(): MeditationTimer {
        return copy(lastIntervalGongAt = remainingSeconds)
    }

    /**
     * Checks if an interval gong should be played.
     *
     * Supports 3 modes via [IntervalMode]:
     * - REPEATING: Gongs at every full interval from elapsed time (5:00, 10:00, 15:00...)
     * - AFTER_START: Single gong X minutes after start
     * - BEFORE_END: Single gong X minutes before end
     *
     * The 5-second protection prevents collision with the end gong.
     *
     * @param intervalMinutes Interval in minutes (1-60)
     * @param mode How interval gongs are triggered
     * @return True if a gong should be played at the current remaining time
     */
    @Suppress("ReturnCount") // Multiple guard clauses for clarity
    fun shouldPlayIntervalGong(intervalMinutes: Int, mode: IntervalMode = IntervalMode.REPEATING): Boolean {
        if (state != TimerState.Running) return false
        if (intervalMinutes <= 0) return false

        val intervalSeconds = intervalMinutes * SECONDS_PER_MINUTE

        // No gong if interval >= total duration
        if (intervalSeconds >= totalSeconds) return false

        // 5-second protection: no gong in final 5 seconds to avoid collision with end gong
        if (remainingSeconds <= END_GONG_PROTECTION_SECONDS) return false

        return when (mode) {
            IntervalMode.REPEATING -> shouldPlayRepeatingFromStart(intervalSeconds)
            IntervalMode.AFTER_START -> shouldPlaySingleFromStart(intervalSeconds)
            IntervalMode.BEFORE_END -> shouldPlaySingleFromEnd(intervalSeconds)
        }
    }

    /**
     * The effective start point for interval calculations.
     * Uses silentPhaseStartRemaining when attunement was played, otherwise totalSeconds.
     */
    private val effectiveStartRemaining: Int
        get() = silentPhaseStartRemaining ?: totalSeconds

    private fun shouldPlaySingleFromEnd(intervalSeconds: Int): Boolean {
        // Already played?
        if (lastIntervalGongAt != null) return false

        return remainingSeconds <= intervalSeconds
    }

    private fun shouldPlaySingleFromStart(intervalSeconds: Int): Boolean {
        // Already played?
        if (lastIntervalGongAt != null) return false

        val elapsed = effectiveStartRemaining - remainingSeconds
        return elapsed >= intervalSeconds
    }

    private fun shouldPlayRepeatingFromStart(intervalSeconds: Int): Boolean {
        // First gong not yet played
        if (lastIntervalGongAt == null) {
            val elapsed = effectiveStartRemaining - remainingSeconds
            return elapsed >= intervalSeconds
        }

        // Check if enough time passed since last gong
        val timeSinceLastGong = lastIntervalGongAt - remainingSeconds
        return timeSinceLastGong >= intervalSeconds
    }

    /** Returns a reset timer with original duration */
    fun reset(): MeditationTimer {
        return copy(
            remainingSeconds = durationMinutes * 60,
            state = TimerState.Idle,
            remainingPreparationSeconds = 0,
            silentPhaseStartRemaining = null,
            lastIntervalGongAt = null
        )
    }

    companion object {
        const val DEFAULT_PREPARATION_TIME = 15
        const val MIN_DURATION_MINUTES = 1
        const val MAX_DURATION_MINUTES = 60
        private const val SECONDS_PER_MINUTE = 60

        /** Protection zone: no interval gong in last 5 seconds to avoid collision with end gong */
        const val END_GONG_PROTECTION_SECONDS = 5

        /**
         * Creates a new meditation timer with validated duration.
         *
         * @param durationMinutes Duration in minutes (1-60)
         * @param preparationTimeSeconds Duration of preparation in seconds (default: 15). Use 0 to skip preparation.
         * @param attunementDurationSeconds Duration of attunement audio in seconds (default: 0 = no attunement)
         * @return A new MeditationTimer instance
         * @throws IllegalArgumentException if duration is not between 1 and 60 minutes
         */
        fun create(
            durationMinutes: Int,
            preparationTimeSeconds: Int = DEFAULT_PREPARATION_TIME,
            attunementDurationSeconds: Int = 0
        ): MeditationTimer {
            return MeditationTimer(
                durationMinutes = durationMinutes,
                remainingSeconds = durationMinutes * 60,
                state = TimerState.Idle,
                remainingPreparationSeconds = 0,
                preparationTimeSeconds = preparationTimeSeconds,
                attunementDurationSeconds = attunementDurationSeconds,
                silentPhaseStartRemaining = null,
                lastIntervalGongAt = null
            )
        }
    }
}

File: ios/StillMoment/Domain/Models/MeditationTimer.swift
//
//  MeditationTimer.swift
//  Still Moment
//
//  Domain Model - Meditation Timer
//

import Foundation

/// Errors that can occur when creating a meditation timer
enum MeditationTimerError: Error, LocalizedError {
    case invalidDuration(Int)

    // MARK: Internal

    var errorDescription: String? {
        switch self {
        case let .invalidDuration(minutes):
            "Invalid duration: \(minutes) minutes. Duration must be between 1 and 60 minutes."
        }
    }
}

/// Domain model representing a meditation timer session
///
/// State machine: idle → preparation → startGong → [introduction →] running → endGong → completed
/// Preparation and introduction are optional phases.
/// The introduction phase is event-driven (audio callback), not countdown-driven.
struct MeditationTimer: Equatable {
    // MARK: Lifecycle

    /// Initializes a new meditation timer
    /// - Parameters:
    ///   - durationMinutes: Duration in minutes (1-60)
    ///   - preparationTimeSeconds: Duration of preparation phase in seconds (default: 15). Use 0 to skip.
    /// - Throws: `MeditationTimerError.invalidDuration` if duration is not between 1 and 60 minutes
    init(durationMinutes: Int, preparationTimeSeconds: Int = 15) throws {
        guard (1...60).contains(durationMinutes) else {
            throw MeditationTimerError.invalidDuration(durationMinutes)
        }
        self.durationMinutes = durationMinutes
        self.remainingSeconds = durationMinutes * 60
        self.state = .idle
        self.remainingPreparationSeconds = 0
        self.preparationTimeSeconds = preparationTimeSeconds
        self.silentPhaseStartRemaining = nil
        self.lastIntervalGongAt = nil
    }

    /// Private initializer for internal state updates
    private init(
        durationMinutes: Int,
        remainingSeconds: Int,
        state: TimerState,
        remainingPreparationSeconds: Int = 0,
        preparationTimeSeconds: Int,
        silentPhaseStartRemaining: Int? = nil,
        lastIntervalGongAt: Int? = nil
    ) {
        self.durationMinutes = durationMinutes
        self.remainingSeconds = remainingSeconds
        self.state = state
        self.remainingPreparationSeconds = remainingPreparationSeconds
        self.preparationTimeSeconds = preparationTimeSeconds
        self.silentPhaseStartRemaining = silentPhaseStartRemaining
        self.lastIntervalGongAt = lastIntervalGongAt
    }

    // MARK: Internal

    /// Duration of the timer in minutes (1-60)
    let durationMinutes: Int

    /// Remaining time in seconds
    let remainingSeconds: Int

    /// Current state of the timer
    let state: TimerState

    /// Remaining preparation seconds (preparationTimeSeconds→0 before timer starts)
    let remainingPreparationSeconds: Int

    /// Duration of preparation phase in seconds (configured at initialization)
    let preparationTimeSeconds: Int

    /// Remaining seconds when the silent meditation phase started (after introduction)
    /// Used as baseline for interval gong calculations to exclude introduction time
    let silentPhaseStartRemaining: Int?

    /// Remaining seconds when last interval gong was played
    let lastIntervalGongAt: Int?

    /// Returns total duration in seconds
    var totalSeconds: Int {
        self.durationMinutes * 60
    }

    /// Returns progress as a value between 0.0 and 1.0
    var progress: Double {
        guard self.totalSeconds > 0 else {
            return 0.0
        }
        return 1.0 - (Double(self.remainingSeconds) / Double(self.totalSeconds))
    }

    /// Checks if timer has completed
    var isCompleted: Bool {
        self.remainingSeconds <= 0
    }

    /// Returns a copy with updated remaining seconds and any domain events that occurred.
    ///
    /// - Parameter intervalSettings: Optional interval gong configuration. When provided,
    ///   `tick()` checks if an interval gong is due and emits `.intervalGongDue`.
    ///   When `nil`, no interval gong detection occurs.
    /// - Returns: Tuple of (updated timer, events that occurred during this tick)
    func tick(intervalSettings: IntervalSettings? = nil) -> (MeditationTimer, [TimerEvent]) {
        switch self.state {
        case .preparation:
            let newTimer = self.tickPreparation()
            let events: [TimerEvent] = newTimer.state == .startGong ? [.preparationCompleted] : []
            return (newTimer, events)
        case .startGong:
            let newTimer = self.tickRunning()
            let events: [TimerEvent] = newTimer.state == .endGong ? [.meditationCompleted] : []
            return (newTimer, events)
        case .introduction:
            let newTimer = self.tickIntroduction()
            let events: [TimerEvent] = newTimer.state == .endGong ? [.meditationCompleted] : []
            return (newTimer, events)
        case .running:
            return self.tickRunningWithEvents(intervalSettings: intervalSettings)
        case .idle,
             .endGong,
             .completed:
            return (self, [])
        }
    }

    /// Ticks the running state with interval gong detection and event emission
    private func tickRunningWithEvents(
        intervalSettings: IntervalSettings?
    ) -> (MeditationTimer, [TimerEvent]) {
        let ticked = self.tickRunning()

        if ticked.state == .endGong {
            return (ticked, [.meditationCompleted])
        }

        if let settings = intervalSettings,
           ticked.shouldPlayIntervalGong(
               intervalMinutes: settings.intervalMinutes,
               mode: settings.mode
           ) {
            return (ticked.markIntervalGongPlayed(), [.intervalGongDue])
        }

        return (ticked, [])
    }

    /// Ticks the preparation countdown.
    /// Transitions to `.startGong` when preparation finishes.
    private func tickPreparation() -> MeditationTimer {
        let newPreparation = max(0, self.remainingPreparationSeconds - 1)
        let newState: TimerState = newPreparation <= 0 ? .startGong : .preparation
        return MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.remainingSeconds,
            state: newState,
            remainingPreparationSeconds: newPreparation,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.silentPhaseStartRemaining,
            lastIntervalGongAt: self.lastIntervalGongAt
        )
    }

    /// Ticks the introduction phase (meditation timer decrements, never auto-transitions to running).
    /// The transition to `.running` is event-driven via `endIntroduction()`.
    private func tickIntroduction() -> MeditationTimer {
        let newRemaining = max(0, self.remainingSeconds - 1)
        let newState: TimerState = newRemaining <= 0 ? .endGong : .introduction
        return MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: newRemaining,
            state: newState,
            remainingPreparationSeconds: self.remainingPreparationSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.silentPhaseStartRemaining,
            lastIntervalGongAt: self.lastIntervalGongAt
        )
    }

    /// Ticks the main meditation timer (used for .startGong and .running states)
    private func tickRunning() -> MeditationTimer {
        let newRemaining = max(0, self.remainingSeconds - 1)
        let newState: TimerState = newRemaining <= 0 ? .endGong : self.state
        return MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: newRemaining,
            state: newState,
            remainingPreparationSeconds: self.remainingPreparationSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.silentPhaseStartRemaining,
            lastIntervalGongAt: self.lastIntervalGongAt
        )
    }

    /// Returns a copy with updated state
    func withState(_ newState: TimerState) -> MeditationTimer {
        MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.remainingSeconds,
            state: newState,
            remainingPreparationSeconds: self.remainingPreparationSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.silentPhaseStartRemaining,
            lastIntervalGongAt: self.lastIntervalGongAt
        )
    }

    /// Returns a copy ready for preparation phase (uses configured preparationTimeSeconds)
    func startPreparation() -> MeditationTimer {
        MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.remainingSeconds,
            state: .preparation,
            remainingPreparationSeconds: self.preparationTimeSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: nil,
            lastIntervalGongAt: nil
        )
    }

    /// Returns a copy transitioned from `.introduction` to `.running`.
    /// Called when the introduction audio finishes playing (event-driven).
    /// Sets `silentPhaseStartRemaining` to current remaining seconds for interval gong calculations.
    func endIntroduction() -> MeditationTimer {
        MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.remainingSeconds,
            state: .running,
            remainingPreparationSeconds: self.remainingPreparationSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.remainingSeconds,
            lastIntervalGongAt: self.lastIntervalGongAt
        )
    }

    /// Returns a copy with updated interval gong timestamp
    func markIntervalGongPlayed() -> MeditationTimer {
        MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.remainingSeconds,
            state: self.state,
            remainingPreparationSeconds: self.remainingPreparationSeconds,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: self.silentPhaseStartRemaining,
            lastIntervalGongAt: self.remainingSeconds
        )
    }

    /// Checks if an interval gong should be played based on the interval mode
    /// - Parameters:
    ///   - intervalMinutes: Interval in minutes (e.g., 5 for every 5 minutes)
    ///   - mode: The interval mode (repeating, afterStart, beforeEnd)
    /// - Returns: True if gong should be played now
    func shouldPlayIntervalGong(intervalMinutes: Int, mode: IntervalMode = .repeating) -> Bool {
        guard self.state == .running else {
            return false
        }
        guard intervalMinutes > 0 else {
            return false
        }

        // 5-second protection: no gong in the last 5 seconds to avoid collision with end gong
        guard self.remainingSeconds > 5 else {
            return false
        }

        let intervalSeconds = intervalMinutes * 60

        // No gong if interval >= total duration (consistent with Android)
        guard intervalSeconds < self.totalSeconds else {
            return false
        }

        switch mode {
        case .repeating:
            return self.shouldPlayRepeatingGong(intervalSeconds: intervalSeconds)
        case .afterStart:
            return self.shouldPlayAfterStartGong(intervalSeconds: intervalSeconds)
        case .beforeEnd:
            return self.shouldPlayBeforeEndGong(intervalSeconds: intervalSeconds)
        }
    }

    // MARK: - Private Interval Gong Helpers

    /// The effective start point for interval calculations
    /// Uses silentPhaseStartRemaining when introduction was played, otherwise totalSeconds
    private var effectiveStartRemaining: Int {
        self.silentPhaseStartRemaining ?? self.totalSeconds
    }

    /// Repeating mode: gongs at every full interval from start of silent meditation
    private func shouldPlayRepeatingGong(intervalSeconds: Int) -> Bool {
        // Never played before - play if we've passed first interval since silent phase started
        guard let lastGongAt = self.lastIntervalGongAt else {
            let elapsed = self.effectiveStartRemaining - self.remainingSeconds
            return elapsed >= intervalSeconds
        }

        // Check if enough time passed since last gong
        let timeSinceLastGong = lastGongAt - self.remainingSeconds
        return timeSinceLastGong >= intervalSeconds
    }

    /// After start mode: exactly 1 gong X minutes after start of silent meditation
    private func shouldPlayAfterStartGong(intervalSeconds: Int) -> Bool {
        // Already played - single gong only
        guard self.lastIntervalGongAt == nil else {
            return false
        }

        let elapsed = self.effectiveStartRemaining - self.remainingSeconds
        return elapsed >= intervalSeconds
    }

    /// Before end mode: exactly 1 gong X minutes before end
    private func shouldPlayBeforeEndGong(intervalSeconds: Int) -> Bool {
        // Already played - single gong only
        guard self.lastIntervalGongAt == nil else {
            return false
        }

        return self.remainingSeconds <= intervalSeconds
    }

    /// Returns a reset timer with original duration
    ///
    /// Uses the private initializer since durationMinutes is already validated
    func reset() -> MeditationTimer {
        MeditationTimer(
            durationMinutes: self.durationMinutes,
            remainingSeconds: self.durationMinutes * 60,
            state: .idle,
            remainingPreparationSeconds: 0,
            preparationTimeSeconds: self.preparationTimeSeconds,
            silentPhaseStartRemaining: nil,
            lastIntervalGongAt: nil
        )
    }
}
